base/strings/iteration.jl

```diff
@@ -0,0 +1,120 @@
+# A specialized iterator for EachIndex of strings
+struct EachStringIndex{T<:AbstractString}
+    s::T
+end
+keys(s::AbstractString) = EachStringIndex(s)
+
+length(e::EachStringIndex) = length(e.s)
+first(::EachStringIndex) = 1
+last(e::EachStringIndex) = lastindex(e.s)
+eltype(::Type{<:EachStringIndex}) = Int
+
+# Iteration over StringNext
+#
+# Any new subtype of AbstractString, should override
+#
+#   next(::StringNext{MyString}, state)
+#
+# to provide iteration over the string and its indices. All other iteration methods,
+# including iteration over strings, iteration over pairs, indexing into string,
+# iteration over indicies alone are derived from this method.
+
+const StringNext{T<:AbstractString} = Iterators.Next{T, EachStringIndex{T}}
+StringNext(x::T) where {T<:AbstractString} = Next(x)
+StringNext(x::T, idx) where {T<:AbstractString} = Next(x, idx)
+StringNext(x::T, idx, itr) where {T<:AbstractString} = Next(x, idx, itr)
+
+start(sp::StringNext) = 1
```

Comment:
Should it be `start(sp::StringNext) = firstindex(sp.data)` ?
The doc says that string indexing starts at `1`, but there are [other places that use `firstindex`](https://github.com/JuliaLang/julia/blob/master/base/strings/search.jl#L102) _(same for `first(::EachStringIndex) = 1` above)_.

The more places the `1` is not hard coded, the easier it'll be to eventually support `AbstractString` types with non standard indexing.

Reply:
Part of the goal of this change is to separate the assumption currently baked into the string code that indices and itartion state of strings are the same thing. We want indices to _always_ be `1` through `ncodeunits(s)` but allow iteration state to be anything at all (even non-integers). So `firstindex(s)` and `start(s)` should be totally unrelated: the former is always `1` but the latter can be anything at all.

Reply:
`start(s)` here is just a default (custom string types can override it). It is also going away with the iteration protocol change. I do want to eventually define non-standard indexing for strings, but that's a natural extension of this and string should always support linear indicies.

Reply:
> We want indices to always be 1 through ncodeunits(s)

I've been trying out various string implementations as a way to understand what the abstract interface contract is. If `firstindex(s) == 1` is part of the protocol, so be it.

Is it also to be assumed that `isvalid(s, firstindex(s)) == true`? If this wasn't required that would allow "padding" of the codeunit space in cases where that is more efficient.

I have a `SplicedString` implementation where the indexes are in the range `1:ncodeunits(s)` (and `isvalid(s, 1) == true`), but where the index space is only sparsely populated.
It all works nicely for constant time indexing and iteration, but anyone who writes a `i = i+1` loop is going to find it take a long time to run. Is this too heretical?

```julia
julia> s = LazyJSON.SplicedString("Foo", " ", "Bar")
"Foo Bar"

julia> [keys(s)...]
7-element Array{Int64,1}:
             1
             2
             3
 1099511627777
 2199023255553
 2199023255554
 2199023255555

julia> [(i >> 40, i & (2^40-1)) for i in keys(s)]
7-element Array{Tuple{Int64,Int64},1}:
 (0, 1)
 (0, 2)
 (0, 3)
 (1, 1)
 (2, 1)
 (2, 2)
 (2, 3)
```

Reply:
> firstindex(s) == 1

I'd be fine relaxing this constraint, but I don't care too much.

> isvalid(s, firstindex(s)) == true

This one I do care about.

> Is this too heretical?

I mean, you can do various kinds of encoding here, this one included, but I'd rather go the direction where strings can have non-standard index kinds, but all still need to support linear indexing no matter how slow it is.